test/parallel/test-crypto-fips.js

```diff
@@ -0,0 +1,208 @@
+const CNF_FIPS_ON = path.join(common.fixturesDir, 'openssl_fips_enabled.cnf');
+const CNF_FIPS_OFF = path.join(common.fixturesDir, 'openssl_fips_disabled.cnf');
+var num_children_spawned = 0;
+var num_children_ok = 0;
+
+function compiledWithFips() {
+  return process.config.variables.openssl_fips ? true : false;
+}
+
+function addToEnv(newVar, value) {
+  var envCopy = {};
+  for (const e in process.env) {
+    envCopy[e] = process.env[e];
+  }
+  envCopy[newVar] = value;
+  return envCopy;
+}
+
+function childOk(child) {
+  console.error('Child ' +  ++num_children_ok + '/' + num_children_spawned +
+      ' [pid:' + child.pid + '] OK.');
+}
+
+function testHelper(requiresFips, args, expectedOutput, cmd, env) {
+  const child = spawn(process.execPath, args.concat(['-i']), {
+    cwd: path.dirname(process.execPath),
+    env: env
+  });
+
+  console.error('Spawned child [pid:' + child.pid + '] with cmd ' +
+      cmd + ' and args \'' + args + '\'');
+  num_children_spawned++;
+
+  function response_handler(response, expectedOutput, expectedError) {
+    // We won't always have data on both stdout and stderr.
+    if (response.length > 0) {
+      if (EXIT_FAILURE === expectedOutput) {
+        if (-1 === response.indexOf(expectedError))
+          assert.notEqual(-1, response.indexOf(expectedError));
+      } else {
+        assert.equal(expectedOutput, response);
+      }
+      childOk(child);
+    }
+  }
+
+  // Buffer all data received from children on stderr/stdout.
+  const buffer = { stderr: '', stdout: '' };
+  const response_buffer = function(stream, data) {
+    // Prompt and newline may occur in undefined order.
+    const response = data.toString().replace(/\n|>/g, '').trim();
+    if (response.length > 0) {
+      buffer[stream] += response;
+    }
+  };
+
+  child.stdout.on('data', function(data) {
+    response_buffer('stdout', data);
+  });
+
+  child.stderr.on('data', function(data) {
+    response_buffer('stderr', data);
+  });
+
+  // If using FIPS mode binary, or running a generic test.
+  if (compiledWithFips() || !requiresFips) {
+    child.stdout.on('end', function(data) {
+      response_handler(buffer['stdout'],
+          expectedOutput, FIPS_ERROR_STRING);
+    });
+  } else {
+    // If using a non-FIPS binary expect a failure.
+    child.stdout.on('end', function() {
+      response_handler(buffer['stdout'],
+          EXIT_FAILURE, FIPS_ERROR_STRING);
+    });
+    /* Failure to start Node executable is reported on stderr */
+    child.stderr.on('end', function() {
+      response_handler(buffer['stderr'],
+          EXIT_FAILURE, OPTION_ERROR_STRING);
+    });
+  }
+  // Wait for write to complete before exiting child, but we won't be able to
+  // write to the child's stdin if cmd line args are rejected and process dies
+  // first, resulting in EPIPE. We can avoid EPIPE by  not writing anything.
+  if (null !== cmd) {
+    child.stdin.setEncoding('utf-8');
+    child.stdin.write(cmd + '\n;process.exit()\n');
```

Comment:
Just curious... could we possibly start child process with `-e` command-line argument instead? And maybe use `child_process.spawnSync` to simplify all of the code above?